scanner/api/sink.h

```diff
@@ -0,0 +1,181 @@
+/* Copyright 2017 Carnegie Mellon University
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "scanner/api/kernel.h"
+#include "scanner/util/common.h"
+#include "scanner/util/profiler.h"
+
+#include <vector>
+
+namespace scanner {
+
+//! Parameters provided at instantiation of Sink node
+struct SinkConfig {
+  std::vector<std::string> input_columns;
+  std::vector<proto::ColumnType> input_column_types;
+  std::vector<u8> args;  //! Byte-string of proto args if given.
+  i32 node_id;
+};
+
+/**
+ * @brief Interface for reading data in a computation graph.
+ *
+ * Sink are how Scanner's computation graphs read data from outside the
+ * system.
+ */
+class Sink {
+ public:
+  Sink(const SinkConfig& config) {}
+
+  virtual ~Sink(){};
+
+  /**
+   * @brief Checks if Sink arguments are valid.
+   *
+   * Only useful if your Sink has its own custom Protobuf arguments.
+   */
+  virtual void validate(proto::Result* result) { result->set_success(true); }
+
+  /**
+   * @brief Informs the Sink that it will now be processing an new stream.
+   *
+   * @param args
+   *        the arguments that were bound to this output for this stream
+   *
+   */
+  virtual void new_stream(const std::vector<u8>& args) = 0;
+
+  /**
+   * @brief Runs the Sink to write elements.
+   *
+   * @param input_columns
+   *        sink input, vector of elements produced by an Op
+   *
+   */
+  virtual void write(const BatchedElements& input_columns) = 0;
+
+  /**
+   * @brief Called when the Sink has received the end of a contiguous sequence
+   *        of data items (for example, when the last element of a stream is
+   *        provided to the Sink).
+   *
+   */
+  virtual void finished() {};
+
+  /**
+   * @brief For internal use
+   **/
+  virtual void set_profiler(Profiler* profiler) { profiler_ = profiler; }
```

Comment:
There's only ever one profiler per process right? We could just make this a global instead of passing pointers everywhere.

Reply:
No, there are multiple. Currently there is one per thread of control (evaluate worker, save worker, load worker, etc). In general, I hate using globals since it makes it much harder to understand where the dependencies are. State should be passed explicitly.